libraries/pod/src/pod_option.rs

```diff
@@ -0,0 +1,218 @@
+//! Generic `Option` that can be used as a `Pod`s for types that can have
+//! a `None` value.
+
+use {
+    bytemuck::{Pod, Zeroable},
+    solana_program::{program_option::COption, pubkey::Pubkey},
+};
+
+/// Trait for types that can be `None`.
+///
+/// This trait is used to indicate that a type can be `None` according to a
+/// specific value.
+pub trait Nullable: Pod {
+    /// Indicates whether the value is `None` or not.
+    fn is_none(&self) -> bool;
+}
+
+/// A "pod-enabled" type that can be used as an `Option<T>` without
+/// requiring extra space to indicate if the value is `Some` or `None`.
+///
+/// This can be used when a specific value of `T` indicates that its
+/// value is `None`.
+#[repr(transparent)]
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub struct PodOption<T: Nullable>(T);
+
+unsafe impl<T: Nullable> Pod for PodOption<T> {}
+
+unsafe impl<T: Nullable> Zeroable for PodOption<T> {}
+
+impl<T: Nullable> PodOption<T> {
+    #[inline]
+    pub fn new(value: T) -> Self {
+        Self(value)
+    }
+
+    #[inline]
+    pub fn to_option(&self) -> Option<&T> {
+        if self.0.is_none() {
+            None
+        } else {
+            Some(&self.0)
+        }
+    }
+
+    #[inline]
+    pub fn to_option_mut(&mut self) -> Option<&mut T> {
+        if self.0.is_none() {
+            None
+        } else {
+            Some(&mut self.0)
+        }
+    }
+}
+
+impl<T: Nullable> From<PodOption<T>> for Option<T> {
+    fn from(value: PodOption<T>) -> Self {
+        if value.0.is_none() {
+            None
+        } else {
+            Some(value.0)
+        }
+    }
+}
+
+impl<T: Nullable> From<PodOption<T>> for COption<T> {
+    fn from(value: PodOption<T>) -> Self {
+        if value.0.is_none() {
+            COption::None
+        } else {
+            COption::Some(value.0)
+        }
+    }
+}
+
+/// Simple macro for implementing the `Nullable` trait on int types.
+///
+/// The implementation assumes that the value `0` represents `None`.
+macro_rules! impl_int_nullable {
+    ( $I:ty ) => {
+        impl Nullable for $I {
+            fn is_none(&self) -> bool {
+                *self == 0
+            }
+        }
+    };
+}
+
+impl_int_nullable!(u8);
+impl_int_nullable!(u16);
+impl_int_nullable!(u32);
+impl_int_nullable!(u64);
+impl_int_nullable!(u128);
```

Comment:
I've only just found all of these implementations in bytemuck already -- maybe we can use them instead? https://docs.rs/bytemuck/latest/bytemuck/trait.PodInOption.html

Reply:
Oh, this is interesting. We could use them. The only comment that I have is that we will have two "types" of optionals:
* `PodOption<T>`: when T is not one of the `NonZero*` variants
* `Option<T>`: when T is one of the `NonZero*` variants

We could still `impl` the `Nullable` trait on the `NonZero*` so you could use `PodOption` with them as well. It does not prevent someone from using `Option<NonZero*>` but gives the alternative to have all `PodOption` instead. 

In summary, we don't `impl` the `Nullable` trait on primitives, only on the `NonZero*` variants. If you want to have a `PodOption` with a primitive value, you need to use the `NonZero*` variant, which makes it explicit that zero is not a valid value.

Reply:
To be honest, I typically prefer less code, so I would recommend using `Option<NonZero*>` over potentially confusing people with two ways of doing things.

Reply:
Good point!